src/lib/transition_frontier/persistent_frontier/persistent_frontier.ml

```diff
@@ -0,0 +1,290 @@
+    Deferred.return
+      (Staged_ledger.Scan_state.staged_transactions root.scan_state)
+  in
+  let mask = Ledger.of_database root_ledger in
+  let%bind () =
+    Deferred.return
+      (List.fold transactions ~init:(Or_error.return ()) ~f:(fun acc txn ->
+           let open Or_error.Let_syntax in
+           let%bind () = acc in
+           let%map _ = Ledger.apply_transaction mask txn in
+           () ))
+  in
+  Staged_ledger.of_scan_state_and_ledger ~logger ~verifier ~snarked_ledger_hash
+    ~ledger:mask ~scan_state:root.scan_state
+    ~pending_coinbase_collection:root.pending_coinbase
+
+module rec Instance_type : sig
+  type t =
+    {db: Database.t; mutable sync: Sync.t option; factory: Factory_type.t}
+end =
+  Instance_type
+
+and Factory_type : sig
+  type t =
+    { logger: Logger.t
+    ; directory: string
+    ; verifier: Verifier.t
+    ; time_controller: Block_time.Controller.t
+    ; mutable instance: Instance_type.t option }
+end =
+  Factory_type
+
+open Instance_type
+open Factory_type
+
+module Instance = struct
+  type t = Instance_type.t
+
+  let create factory =
+    let db =
+      Database.create ~logger:factory.logger ~directory:factory.directory
+    in
+    {db; sync= None; factory}
+
+  let assert_no_sync t =
+    if Option.is_some t.sync then Error `Sync_cannot_be_running else Ok ()
+
+  let assert_sync t ~f =
+    match t.sync with
+    | None ->
+        return (Error `Sync_must_be_running)
+    | Some sync ->
+        f sync
+
+  let start_sync t =
+    let open Result.Let_syntax in
+    let%bind () = assert_no_sync t in
+    let%map base_hash = Database.get_frontier_hash t.db in
+    t.sync
+    <- Some
+         (Sync.create ~logger:t.factory.logger
+            ~time_controller:t.factory.time_controller ~base_hash ~db:t.db)
+
+  let stop_sync t =
+    let open Deferred.Let_syntax in
+    assert_sync t ~f:(fun sync ->
+        let%map () = Sync.close sync in
+        t.sync <- None ;
+        Ok () )
+
+  let notify_sync t ~diffs ~hash_transition =
+    assert_sync t ~f:(fun sync ->
+        Sync.notify sync ~diffs ~hash_transition ;
+        Deferred.Result.return () )
+
+  let destroy t =
+    let open Deferred.Let_syntax in
+    Logger.trace t.factory.logger ~module_:__MODULE__ ~location:__LOC__
+      "Destroying transition frontier persistence instance" ;
+    let%map () =
+      if Option.is_some t.sync then
+        stop_sync t
+        >>| Fn.compose Result.ok_or_failwith
+              (Result.map_error ~f:(Fn.const "impossible"))
+      else return ()
+    in
+    Database.close t.db ;
+    t.factory.instance <- None
+
+  let factory {factory; _} = factory
+
+  let check_database t = Database.check t.db
+
+  let get_root_transition t =
+    let open Result.Let_syntax in
+    Database.get_root_hash t.db
+    >>= Database.get_transition t.db
+    |> Result.map_error ~f:Database.Error.message
+
+  let set_frontier_hash t = Database.set_frontier_hash t.db
+
+  let fast_forward t target_root :
```

Comment:
don't like the name of this function. When I think of fast_forward, I think of loading the persistent_frontier. However, it just seems to check the root_idneitifer. Can you change the name of the function?

Reply:
This is step 1 of fast forward. Fast forward here refers to fast forwarding the root, but there is some extra work long term that needs to be implemented before we can actually do that (discussed this with Corey, will write up an issue and work on it once this is closed). So, this is like a stub for fast forward since right now it can succeed if the root doesn't need to be moved. This function will contain much more logic in the next PR.